API and Microservice Architecture.md

```diff
@@ -0,0 +1,72 @@
+1. **What is an API? Explain and give examples of different types and implementations.**
+    - **API (Application Programming Interface):** An API is a set of rules that allows one piece of software to interact with another. It defines methods and data structures for such interactions.
+    - **Types of APIs:**
+        - **REST APIs:** Used for web services, REST (Representational State Transfer) APIs allow interaction over HTTP with resources identified by URLs. E.g., Twitter API.
+        - **SOAP APIs:** Simple Object Access Protocol is a protocol used for exchanging structured information in web services. E.g., PayPal API.
+        - **GraphQL APIs:** A query language for your API that allows clients to request exactly the data they need. E.g., GitHub API.
+        - **Operating System APIs:** Interfaces to interact with the operating system. E.g., Windows API.
+  
+2. **What is an SDK?**
+    - **SDK (Software Development Kit):** An SDK is a collection of software development tools, libraries, documentation, and sample code that developers use to build applications for specific platforms or environments. 
+  
+3. **What is a Microservice Architecture?**
+    - **Microservice Architecture:** It is an architectural style that structures an application as a collection of loosely coupled, independently deployable services. Each microservice is responsible for a specific business function and communicates with other microservices via APIs.
+  
+4. **Explain how the architecture works, add a cool diagram.**
+    - **How it works:** In a microservice architecture, each service is a small, self-contained unit that performs a single function. Services communicate with each other over a network, typically using HTTP/REST or messaging queues. They can be developed, deployed, and scaled independently.
+    - **Diagram:**  In a separate file.
+  
+5. **What are the pros and cons of this architecture?**
+    - **Pros:**
+        - Independent deployment and scalability.
+        - Improved fault isolation.
+        - Flexibility in technology stack.
+        - Easier to develop and maintain smaller, focused teams.
```

Comment:
Don't worry when I finish the Hafifa I will try to be as focused as possible:)
teams are more focused because each team is typically responsible for a specific microservice that corresponds to a particular business function or domain. This allows teams to specialize and become experts in the functionality and technology stack of their assigned microservice, leading to greater efficiency and ownership. Additionally, since each microservice is developed, deployed, and maintained independently, teams can work autonomously without being tightly coupled to other teams' work, reducing coordination overhead and enabling faster development cycles.